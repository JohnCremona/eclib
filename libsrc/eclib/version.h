// version.h: declaration of functions show_version(), eclib_version()
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
//

#if     !defined(_ECLIB_VERSION_H)
#define _ECLIB_VERSION_H      1       //flags that this file has been included

#include <string>
#include <eclib/templates.h>

// Return the current eclib version as a string, e.g. 'v20210317'

string eclib_version();

// Return the current eclib version as a triple [y,m,d], e.g. string, e.g. [2021,3,17]

vector<int> eclib_date();

// Display the current eclib version and compilation information

void show_version(ostream& os = cout);

// compare current eclib version date with a triple (y,m,d), returning
// +1 if the current version is later (more recent) than (y,m,d), 0 if
// equal and -1 if earlier (older).

int compare_eclib_version(int y, int m, int d);

#endif
