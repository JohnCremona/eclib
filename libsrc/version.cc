// version.cc: implementation of function show_version()
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2012 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include <iostream>
#include <eclib/templates.h>


void show_version()
{
  cerr << "Version compiled on " << __DATE__ << " at " << __TIME__ << " by GCC " << __VERSION__ << "\n";
#ifdef MPFP
  cerr << "using NTL bigints and NTL real and complex multiprecision floating point";
#else
  cerr << "using NTL bigints but no multiprecision floating point";
#endif
  cerr<<endl;
}

