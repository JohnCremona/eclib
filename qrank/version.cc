// version.cc: implementation of function show_version()
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2005 John Cremona
// 
// This file is part of the mwrank package.
// 
// mwrank is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// mwrank is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with mwrank; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
// 
//////////////////////////////////////////////////////////////////////////
 
#include <iostream>

using namespace std;

void show_version()
{
  cerr << "Version compiled on " << __DATE__ << " at " << __TIME__ << " by GCC " << __VERSION__ << "\n";
  cerr << "using base arithmetic option ";
#ifdef LiDIA_ALL
    cerr << "LiDIA_ALL (LiDIA bigints and multiprecision floating point)";
#else
#ifdef LiDIA_INTS
  cerr << "LiDIA_INTS (LiDIA bigints and no multiprecision floating point)";
#else
#ifdef NTL_ALL
  cerr << "NTL_ALL (NTL bigints and multiprecision floating point)";
#else
#ifdef NTL_INTS
  cerr << "NTL_INTS (NTL bigints and no multiprecision floating point)";
#else
  cerr << "libg++, no multiprecision floating point";
#endif
#endif
#endif
#endif
  cerr<<endl;
}

