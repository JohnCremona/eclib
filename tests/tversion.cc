// FILE tversion.cc : Simple test of eclib versioning utilities
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2012 Marcus Mo
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////

#include <iostream>
#include <eclib/version.h>

int main( int argc, char **argv  )
{
  cout << "This eclib version is " << eclib_version() << endl;
  vector<int> date = eclib_date();
  cout << "[ year month day ] = " << date << endl;

  cout << "\nConfigure options and compilation date:\n" << endl;
  show_version();

  cout << "\nTesting comparison of version date and various (year, month, day) date triples:\n" << endl;
  // test version date comparison
  int y, m, d, s;
  y=2020; m=1; d=1;
  s = compare_eclib_version(y,m,d);
  cout << "Date ";
  cout << setw(4)<<std::setfill('0')<<y;
  cout << setw(2)<<std::setfill('0')<<m;
  cout << setw(2)<<std::setfill('0')<<d;
  cout << " is ";
  if (s>0) cout << "before"; else
    if (s<0) cout << "after"; else
      cout << "the same as";
  cout << " the version date ";
  cout << setw(4)<<std::setfill('0')<<date[0];
  cout << setw(2)<<std::setfill('0')<<date[1];
  cout << setw(2)<<std::setfill('0')<<date[2];
  cout << endl;

  y=2030; m=6; d=11;
  s = compare_eclib_version(y,m,d);
  cout << "Date ";
  cout << setw(4)<<std::setfill('0')<<y;
  cout << setw(2)<<std::setfill('0')<<m;
  cout << setw(2)<<std::setfill('0')<<d;
  cout << " is ";
  if (s>0) cout << "before"; else
    if (s<0) cout << "after"; else
      cout << "the same as";
  cout << " the version date ";
  cout << setw(4)<<std::setfill('0')<<date[0];
  cout << setw(2)<<std::setfill('0')<<date[1];
  cout << setw(2)<<std::setfill('0')<<date[2];
  cout << endl;

  y=2021; m=3; d=17;
  s = compare_eclib_version(y,m,d);
  cout << "Date ";
  cout << setw(4)<<std::setfill('0')<<y;
  cout << setw(2)<<std::setfill('0')<<m;
  cout << setw(2)<<std::setfill('0')<<d;
  cout << " is ";
  if (s>0) cout << "before"; else
    if (s<0) cout << "after"; else
      cout << "the same as";
  cout << " the version date ";
  cout << setw(4)<<std::setfill('0')<<date[0];
  cout << setw(2)<<std::setfill('0')<<date[1];
  cout << setw(2)<<std::setfill('0')<<date[2];
  cout << endl;

  exit(0);
}
