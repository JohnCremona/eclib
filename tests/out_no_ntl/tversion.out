This eclib version is 20220621
[ year month day ] = [ 2022 6 21 ]

Configure options and compilation date:

eclib version 20220621, using NTL bigints but no multiprecision floating point

Testing comparison of version date and various (year, month, day) date triples:

Date 20200101 is before the version date 20220621
Date 20300611 is after the version date 20220621
Date 20210317 is before the version date 20220621
